Source/WebCore/css/StyleProperties.cpp

```diff
@@ -488,78 +506,32 @@ void StyleProperties::appendFontLonghandValueIfExplicit(CSSPropertyID propertyID
-    auto lineHeightPropertyIndex = findPropertyIndex(CSSPropertyLineHeight);
-
-    if (sizePropertyIndex == -1
-        || familyPropertyIndex == -1
-        || stylePropertyIndex == -1
-        || variantCapsPropertyIndex == -1
-        || weightPropertyIndex == -1
-        || lineHeightPropertyIndex == -1)
-        return std::nullopt;
-
-    auto sizeProperty = propertyAt(sizePropertyIndex);
-    auto familyProperty = propertyAt(familyPropertyIndex);
-    auto styleProperty = propertyAt(stylePropertyIndex);
-    auto variantCapsProperty = propertyAt(variantCapsPropertyIndex);
-    auto weightProperty = propertyAt(weightPropertyIndex);
-    auto lineHeightProperty = propertyAt(lineHeightPropertyIndex);
-
-    if (sizeProperty.isImplicit()
-        || familyProperty.isImplicit()
-        || styleProperty.isImplicit()
-        || variantCapsProperty.isImplicit()
-        || weightProperty.isImplicit()
-        || lineHeightProperty.isImplicit())
-        return std::nullopt;
-
-    auto* sizeValue = sizeProperty.value();
-    auto* familyValue = familyProperty.value();
-    auto* styleValue = styleProperty.value();
-    auto* variantCapsValue = variantCapsProperty.value();
-    auto* weightValue = weightProperty.value();
-    auto* lineHeightValue = lineHeightProperty.value();
-
-    if (!is<CSSPrimitiveValue>(sizeValue)
-        || !is<CSSPrimitiveValue>(familyValue)
-        || !is<CSSPrimitiveValue>(styleValue)
-        || !is<CSSPrimitiveValue>(variantCapsValue)
-        || !is<CSSPrimitiveValue>(weightValue)
-        || !is<CSSPrimitiveValue>(lineHeightValue))
-        return std::nullopt;
-
-    auto& sizePrimitiveValue = downcast<CSSPrimitiveValue>(*sizeValue);
-    auto& familyPrimitiveValue = downcast<CSSPrimitiveValue>(*familyValue);
-    auto& stylePrimitiveValue = downcast<CSSPrimitiveValue>(*styleValue);
-    auto& variantCapsPrimitiveValue = downcast<CSSPrimitiveValue>(*variantCapsValue);
-    auto& weightPrimitiveValue = downcast<CSSPrimitiveValue>(*weightValue);
-    auto& lineHeightPrimitiveValue = downcast<CSSPrimitiveValue>(*lineHeightValue);
-
-    auto sizeValueID = sizePrimitiveValue.valueID();
-    auto familyValueID = familyPrimitiveValue.valueID();
-    auto styleValueID = stylePrimitiveValue.valueID();
-    auto variantCapsValueID = variantCapsPrimitiveValue.valueID();
-    auto weightValueID = weightPrimitiveValue.valueID();
-    auto lineHeightValueID = lineHeightPrimitiveValue.valueID();
-
-    if (sizeValueID != familyValueID
-        || sizeValueID != styleValueID
-        || sizeValueID != variantCapsValueID
-        || sizeValueID != weightValueID
-        || sizeValueID != lineHeightValueID)
-        return std::nullopt;
-
-    if (sizeValueID == CSSValueInvalid)
-        return std::nullopt;
+    // This is different from the list of longhands in fontShorthand().
+    // This is the set of longhands that are *set* by the font shorthand, not including the many others that are reset to their initial values.
```

Comment:
Is this kind of set something we need for any other shorthands? If so, we consider adding it to the data in CSSProperties.json, and producing some constexpr data / functions for it.

Reply:
The font shorthand has a special consideration, where it resets the values of many longhands to initial values that can’t even be specified in the font shorthand. So I’m not sure how much there will be a repeating pattern.

One thing I did differently in the revised version of this pull request I am currently preparing is that I put this list of properties and values into a header because it’s now used in two places. It would be fine to later refactor so it is generated; would not be a big change.

It’s possible we may need to know the initial values for other shorthands; I am not sure.